### ISSUE TITLE
chef-dk generates broken cookbooks by default

### ISSUE THREAD

**Issue by bahamat** [chef/chef#5650]
## Description

Using chef-dk and `chef-solo` or `chef-local -z`, cookbooks are broken by default.

## Chef Version

12.17.44

## Platform Version

SunOS 5.11 (using the sunos-5.10 package to standardize across the environment)

## Replication Case

1. Install `chef-12.17.44-1.i386.solaris`
2. Content of `/var/chef/solo.rb`

        file_cache_path "/var/chef"
        cookbook_path "/var/chef/cookbooks"
        chef_repo_path "/var/chef"
        json_attribs "/var/chef/test.json"

3. Content of `/var/chef/test.json`

        {
            "recipes": [ "test1" ]
        }

4. Run `mkdir -p /var/chef/cookbooks/; cd /var/chef/cookbooks`
5. Run `chef generate cookbook test1; cd ~`
6. Run `chef-solo --config /var/chef/solo.rb` or `chef-client -z --config /var/chef/solo.rb`

## Client Output

```
================================================================================
Error Syncing Cookbooks:
================================================================================

Unexpected Error:
-----------------
NoMethodError: private method `select' called for nil:NilClass

Platform:
---------
i386-solaris2.10


Running handlers:
[2016-12-14T20:23:18+00:00] ERROR: Running exception handlers
Running handlers complete
[2016-12-14T20:23:18+00:00] ERROR: Exception handlers complete
Chef Client failed. 0 resources updated in 17 seconds
[2016-12-14T20:23:18+00:00] FATAL: Stacktrace dumped to /var/chef/chef-stacktrace.out
[2016-12-14T20:23:18+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2016-12-14T20:23:18+00:00] ERROR: private method `select' called for nil:NilClass
[2016-12-14T20:23:18+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

## Stacktrace

```
Generated at 2016-12-14 20:23:18 +0000
NoMethodError: private method `select' called for nil:NilClass
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/cookbook/synchronizer.rb:205:in `block in remove_deleted_files'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/cookbook/synchronizer.rb:199:in `each'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/cookbook/synchronizer.rb:199:in `remove_deleted_files'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/cookbook/synchronizer.rb:226:in `clear_obsoleted_cookbooks'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/cookbook/synchronizer.rb:147:in `sync_cookbooks'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/policy_builder/expand_node_object.rb:223:in `sync_cookbooks'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/policy_builder/expand_node_object.rb:83:in `setup_run_context'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/client.rb:511:in `setup_run_context'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/client.rb:281:in `run'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/application.rb:294:in `block in fork_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/application.rb:282:in `fork'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/application.rb:282:in `fork_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/application.rb:247:in `block in run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/local_mode.rb:44:in `with_server_connectivity'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/application.rb:235:in `run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/application/client.rb:464:in `sleep_then_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/application/client.rb:451:in `block in interval_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/application/client.rb:450:in `loop'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/application/client.rb:450:in `interval_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/application/client.rb:434:in `run_application'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/application.rb:59:in `run'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/lib/chef/application/solo.rb:226:in `run'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.17.44/bin/chef-solo:25:in `<top (required)>'
/opt/chef/bin/chef-solo:57:in `load'
```

### NOTE: CHEF CLIENT BUGS ONLY

This issue tracker is for the code contained within this repo -- `chef-client`, base `knife` functionality (not
plugins), `chef-apply`, `chef-solo`, `chef-client -z`, etc.

* [Server issues](https://github.com/chef/chef-server/issues/new)
* [ChefDK issues](https://github.com/chef/chef-dk/issues/new)
* Cookbook Issues (see the https://github.com/chef-cookbooks repos or search [Supermarket](https://supermarket.chef.io) or GitHub/Google)

**Comment by coderanger** [comment]
So this issue appears to be [here](https://github.com/chef/chef/blob/89a632571696b53688e86490ceeec77ca00a3197/lib/chef/cookbook/synchronizer.rb#L205) where a segment-level folder doesn't map to an existing segment. I think we just need a `manifest_segment &&` in there?

**Comment by lamont-granquist** [comment]
should we also chefignore the folder in the generator in chef-dk?

**Comment by bahamat** [comment]
FWIW, adding spec and test to chefignore does not work around the error.

**Comment by soundasleep** [comment]
One workaround seems to be deleting the `test/` and `spec/` folders in the generated cookbook. Platform: x86_64-linux (Ubuntu 16.04)

**Comment by schorsch** [comment]
I have no idea why this code is in place but i added a quick and dirty check for such missing manifest_segments because i was getting the same errors
 
https://github.com/chef/chef/blob/89a632571696b53688e86490ceeec77ca00a3197/lib/chef/cookbook/synchronizer.rb#L205 where i prepended:

```ruby
if !manifest_segment       
   Chef::Log.info("Segment not found #{cookbook_name} #{segment}")
elsif manifest_segment.select { |manifest_record| manifest_record["path"] == "#{segment}/#{file}" }.empty?
..
```

My log output points to a lot of missing spec,test,tasks segments like such:
```
INFO: Segment not found yum test
INFO: Segment not found chef-sugar spec
Segment not found compat_resource ci
....
```

# BTW I am about to have enough from chef 

..the public recipes and the whole infrastructure Why?

* every time i updated the chef gems on the server(maybe twice a year) something broke
* every time i update external recipes via submodules something broke
* every time i update external recipes via submodules some other fancy new f** recipe is needed
* I dont want any windows, homebrew or whatever recipes floating around in my private repo just bcs someone thinks installing on all plattforms might be needed
* I dont want to fork and maintain every single public recipe i need as a soluion to the above

So every time i get 'cheffed' again(like today) i think my time in the last 5 years with chef would have better been spent in simple bash scipts for ubuntu/centOS. I mean where is the point of using chef when it eats your precious dev time every time you touch it (twice a year). Yes i don't have to maintain 100' of machines, dev ops is not my job and i just have 10-20 VMs using a local chef-solo dummy solution.

**Comment by vtolstov** [comment]
any news? i'm using latest chef stable and have such errors

**Comment by mrpaws** [comment]
Any word on this?  Killing our current project.

**Comment by akitada** [comment]
I hit this issue when I tried to downgrade `chef-client` from `13.0.118` to `12.20.3` with `chef_client_updater`.

**Comment by akitada** [comment]
I've created a separate issue for the issue mentioned above, with a complete replication case.
https://github.com/chef/chef/issues/6128

**Comment by akitada** [comment]
> file_cache_path "/var/chef"
> cookbook_path "/var/chef/cookbooks"

In this configuration, `/var/chef/cookbooks` is used as a cookbook store AND cookbook cache directory.
Changing `file_cache_path` to `"/var/chef/cache"` should fix the problem.

**Comment by lamont-granquist** [comment]
this should be closed by #6129 / #6128 ?

**Comment by akitada** [comment]
@lamont-granquist I think this was a simple config issue but Chef crash issue should be fixed by #6129.

**Comment by akitada** [comment]
The issue reported here has been fixed and no longer exists. Please close this.
